cpp/src/arrow/compute/kernels/scalar_nested.cc

```diff
@@ -87,6 +89,203 @@ Status GetListElementIndex(const ExecValue& value, T* out) {
+
+    const ArraySpan& list_values = list_array.child_data[0];
+
+    auto list_builder = checked_cast<BuilderType*>(&builder);
+    for (auto i = 0; i < list_array.length; ++i) {
+      const offset_type offset = offsets[i];
+      const offset_type next_offset = offsets[i + 1];
+      if (list_array.IsNull(i)) {
+        RETURN_NOT_OK(list_builder->AppendNull());
+      } else {
+        RETURN_NOT_OK(SetValues<BuilderType>(list_builder, offset, next_offset, &opts,
+                                             &list_values));
+      }
+    }
+    return Status::OK();
+  }
+  template <typename BuilderType>
+  static Status SetValues(BuilderType* list_builder, const offset_type offset,
+                          const offset_type next_offset, const ListSliceOptions* opts,
+                          const ArraySpan* list_values) {
+    auto value_builder = list_builder->value_builder();
+    auto cursor = offset;
+
+    RETURN_NOT_OK(list_builder->Append());
+    while (cursor < offset + (opts->stop.value() - opts->start)) {
+      if (cursor + opts->start >= next_offset) {
+        if constexpr (!std::is_same_v<BuilderType, FixedSizeListBuilder>) {
+          break;  // don't pad nulls for variable sized list output
+        }
+        RETURN_NOT_OK(value_builder->AppendNull());
+      } else {
+        RETURN_NOT_OK(
+            value_builder->AppendArraySlice(*list_values, cursor + opts->start, 1));
+      }
+      ++cursor;
+    }
+    return Status::OK();
+  }
+};
+
+Result<TypeHolder> MakeListSliceResolve(KernelContext* ctx,
+                                        const std::vector<TypeHolder>& types) {
+  const auto& opts = OptionsWrapper<ListSliceOptions>::Get(ctx);
+  const auto list_type = checked_cast<const BaseListType*>(types[0].type);
+  const auto field_name = list_type->field(0)->name();
+  const auto value_type = list_type->field(0)->WithName(field_name);
+  const auto return_fixed_size_list =
+      opts.return_fixed_size_list.value_or(list_type->id() == Type::FIXED_SIZE_LIST);
+  if (return_fixed_size_list) {
+    if (!opts.stop.has_value()) {
+      return Status::NotImplemented(
+          "Unable to produce FixedSizeListArray without `stop` being set.");
+    }
+    return fixed_size_list(value_type,
+                           static_cast<int32_t>(opts.stop.value() - opts.start));
+  } else {
+    // Returning large list if that's what we got in and didn't ask for fixed size
+    if (list_type->id() == Type::LARGE_LIST) {
+      return large_list(value_type);
+    }
+    return list(value_type);
+  }
+}
+
+template <typename InListType, template <typename...> class Functor>
+void AddListSliceKernels(ScalarFunction* func) {
+  for (const auto& index_type : IntTypes()) {
+    auto inputs = {InputType(InListType::type_id)};
+    auto output = OutputType{MakeListSliceResolve};
+    auto scalar_exec = GenerateInteger<Functor, InListType>({index_type->id()});
```

Comment:
I think this "GenerateInteger" is not needed here. It's used for "list_element" kernel, because there the "index" argument is an actual typed kernel input that can be any of the 8 integer types. 
But here, are start/stop/step are options (and have a fixed type), and thus we don't have to generate all variants of the kernel for all integer types.

Reply:
Sharp eye, could also clean up other portions of this now that you brought that up. :broom: 
https://github.com/apache/arrow/pull/14395/commits/00f2f3294b01d3c93872ae182290e12009727875